File: fsharp/FsFind/FsFind/FindSettings.fs
namespace FsFind

open System.Text.RegularExpressions

module FindSettings =
    type t = {
        ArchivesOnly : bool;
        Colorize : bool;
        Debug : bool;
        ExcludeHidden : bool;
        InArchiveExtensions : string list;
        InArchiveFilePatterns : Regex list;
        InDirPatterns : Regex list;
        InExtensions : string list;
        InFilePatterns : Regex list;
        InFileTypes : FileType list;
        IncludeArchives : bool;
        ListDirs : bool;
        ListFiles : bool;
        OutArchiveExtensions : string list;
        OutArchiveFilePatterns : Regex list;
        OutDirPatterns : Regex list;
        OutExtensions : string list;
        OutFilePatterns : Regex list;
        OutFileTypes : FileType list;
        PrintUsage : bool;
        PrintVersion : bool;
        Recursive : bool;
        Paths : string list;
        Verbose : bool
    }

    let DefaultSettings = {
        ArchivesOnly = false;
        Colorize = true;
        Debug = false;
        ExcludeHidden = true;
        InArchiveExtensions = [];
        InArchiveFilePatterns = [];
        InDirPatterns = [];
        InExtensions = [];
        InFilePatterns = [];
        InFileTypes = [];
        IncludeArchives = false;
        ListDirs = false;
        ListFiles = false;
        OutArchiveExtensions = [];
        OutArchiveFilePatterns = [];
        OutDirPatterns = [];
        OutExtensions = [];
        OutFilePatterns = [];
        OutFileTypes = [];
        PrintUsage = false;
        PrintVersion = false;
        Recursive = true;
        Paths = [];
        Verbose = false
    }

    let AddExtensions (exts : string) (extList : string list) : string list =
        List.append extList (FileUtil.ExtensionsListFromString exts)

    let AddPath (path : string) (settings : t) : t =
        { settings with Paths=(List.append settings.Paths [path]) }

    let AddPattern (pattern : string) (patternList : Regex list) : Regex list =
        List.append patternList [Regex(pattern, RegexOptions.Compiled)]

    let SetArchivesOnly (archivesOnly : bool) (settings : t) : t =
        match archivesOnly with
        | true -> { settings with ArchivesOnly=true; IncludeArchives=true }
        | _ -> { settings with ArchivesOnly=false }

    let SetDebug (debug : bool) (settings : t) : t =
        match debug with
        | true -> { settings with Debug=true; Verbose=true }
        | _ -> { settings with Debug=false }

    let FileTypesListToString (lst : FileType list) : string = 
        let rec recListToString (acc : string) (lst : FileType list) =
            match lst with
            | []     -> acc.Trim()
            | [a]    -> (recListToString (acc + " \"" + (FileTypes.ToName a) + "\"") [])
            | h :: t -> (recListToString (acc + " \"" + (FileTypes.ToName h) + "\";") t) in
        sprintf "[%s]" (recListToString "" lst)

    let ToString settings =
        String.concat "" [
            "FindSettings(";
            sprintf "ArchivesOnly: %b" settings.ArchivesOnly;
            sprintf ", Colorize: %b" settings.Colorize;
            sprintf ", Debug: %b" settings.Debug;
            sprintf ", ExcludeHidden: %b" settings.ExcludeHidden;
            sprintf ", InArchiveExtensions: %s" (Common.list_to_string(settings.InArchiveExtensions));
            sprintf ", InArchiveFilePatterns: %s" (Common.list_to_string(settings.InArchiveFilePatterns));
            sprintf ", InDirPatterns: %s" (Common.list_to_string(settings.InDirPatterns));
            sprintf ", InExtensions: %s" (Common.list_to_string(settings.InExtensions));
            sprintf ", InFilePatterns: %s" (Common.list_to_string(settings.InFilePatterns));
            sprintf ", InFileTypes: %s" (FileTypesListToString settings.InFileTypes);
            sprintf ", IncludeArchives: %b" settings.IncludeArchives;
            sprintf ", ListDirs: %b" settings.ListDirs;
            sprintf ", ListFiles: %b" settings.ListFiles;
            sprintf ", OutArchiveExtensions: %s" (Common.list_to_string(settings.OutArchiveExtensions));
            sprintf ", OutArchiveFilePatterns: %s" (Common.list_to_string(settings.OutArchiveFilePatterns));
            sprintf ", OutDirPatterns: %s" (Common.list_to_string(settings.OutDirPatterns));
            sprintf ", OutExtensions: %s" (Common.list_to_string(settings.OutExtensions));
            sprintf ", OutFilePatterns: %s" (Common.list_to_string(settings.OutFilePatterns));
            sprintf ", OutFileTypes: %s" (FileTypesListToString settings.OutFileTypes);
            sprintf ", Paths: %s" (Common.list_to_string(settings.Paths));
            sprintf ", PrintUsage: %b" settings.PrintUsage;
            sprintf ", PrintVersion: %b" settings.PrintVersion;
            sprintf ", Recursive: %b" settings.Recursive;
            sprintf ", Verbose: %b" settings.Verbose;
            ")"
        ]
;;

File: ocaml/mlfind/src/findsettings.ml
open Core.Std
open Common

type t = {
  archivesonly : bool;
  colorize : bool;
  debug : bool;
  excludehidden : bool;
  in_archiveextensions : string list;
  in_archivefilepatterns : Re2.Regex.t list;
  in_dirpatterns : Re2.Regex.t list;
  in_extensions : string list;
  in_filepatterns : Re2.Regex.t list;
  in_filetypes : string list;
  includearchives : bool;
  listdirs : bool;
  listfiles : bool;
  out_archiveextensions : string list;
  out_archivefilepatterns : Re2.Regex.t list;
  out_dirpatterns : Re2.Regex.t list;
  out_extensions : string list;
  out_filepatterns : Re2.Regex.t list;
  out_filetypes : string list;
  paths : string list;
  printusage : bool;
  printversion : bool;
  recursive : bool;
  verbose : bool
}

let default_settings = {
  archivesonly = false;
  colorize = true;
  debug = false;
  excludehidden = true;
  in_archiveextensions = [];
  in_archivefilepatterns = [];
  in_dirpatterns = [];
  in_extensions = [];
  in_filepatterns = [];
  in_filetypes = [];
  in_linesafterpatterns = [];
  in_linesbeforepatterns = [];
  includearchives = false;
  listdirs = false;
  listfiles = false;
  out_archiveextensions = [];
  out_archivefilepatterns = [];
  out_dirpatterns = [];
  out_extensions = [];
  out_filepatterns = [];
  out_filetypes = [];
  paths = [];
  printusage = false;
  printversion = false;
  recursive = true;
  verbose = false
};;

let add_extensions (ext_string : string) (extensions : string list) = 
  let exts = String.split ext_string ~on:(char_of_int 44) in
  List.append extensions exts

let add_filetypes (ft_string : string) (filetypes : string list) = 
  let fts = String.split ft_string ~on:(char_of_int 44) in
  List.append filetypes fts

let set_archivesonly (ss : FindSettings.t) (archivesonly: bool) (ss : FindSettings.t) =
  let includearchives = if archivesonly then archivesonly else ss.includearchives
  { ss with archivesonly=archivesonly; includearchives=includearchives }

let set_debug (ss : FindSettings.t) (debug: bool) (ss : FindSettings.t) =
  let verbose = if debug then debug else ss.verbose
  { ss with debug=debug; verbose=verbose }

let to_string s = 
  String.concat [
    sprintf "{archivesonly=%b" s.archivesonly;
    sprintf "; colorize=%b" s.colorize;
    sprintf "; debug=%b" s.debug;
    sprintf "; excludehidden=%b" s.excludehidden;
    sprintf "; in_archiveextensions=%s" (list_to_string s.in_archiveextensions);
    sprintf "; in_archivefilepatterns=%s" (regexp_list_to_string s.in_archivefilepatterns);
    sprintf "; in_dirpatterns=%s" (regexp_list_to_string s.in_dirpatterns);
    sprintf "; in_extensions=%s" (list_to_string s.in_extensions);
    sprintf "; in_filepatterns=%s" (regexp_list_to_string s.in_filepatterns);
    sprintf "; in_filetypes=%s" (list_to_string s.in_filetypes);
    sprintf "; includearchives=%b" s.includearchives;
    sprintf "; listdirs=%b" s.listdirs;
    sprintf "; listfiles=%b" s.listfiles;
    sprintf "; out_archiveextensions=%s" (list_to_string s.out_archiveextensions);
    sprintf "; out_archivefilepatterns=%s" (regexp_list_to_string s.out_archivefilepatterns);
    sprintf "; out_dirpatterns=%s" (regexp_list_to_string s.out_dirpatterns);
    sprintf "; out_extensions=%s" (list_to_string s.out_extensions);
    sprintf "; out_filepatterns=%s" (regexp_list_to_string s.out_filepatterns);
    sprintf "; out_filetypes=%s" (list_to_string s.out_filetypes);
    sprintf "; printversion=%b" s.printversion;
    sprintf "; recursive=%b" s.recursive;
    sprintf "; paths=%s" (list_to_string s.paths);
    sprintf "; verbose=%b}" s.verbose];;
